File: python/log/log.py
from pydantic import BaseModel

import os
import sys
import inspect
from datetime import datetime

from log.config import config

class Log(BaseModel):

    level : str
    timestamp : datetime
    process : str
    function : str
    file : str
    line : int
    
    def __init__( self, level : str, offset : int = 0 ):
        process = '<cli>'
        if config.pod_name:
            process = config.pod_name
        elif sys.argv[0]:
            process = os.path.basename( sys.argv[ 0 ] )
        stack = inspect.stack()
        super().__init__(
            level = level,
            timestamp = datetime.now(),
            process = process,
            file = os.path.basename( stack[ 2 + offset ][1] ),
            line = stack[ 2 + offset ][2],
            function = stack[ 2 + offset ][3]
        )
                
    def emit( self, message : str ):
        print( f'{self.level} | {self.timestamp} | {self.process} | {self.file}:{self.line} | {self.function} | {message}')
        
def trace( message : str, offset : int = 0 ):
    if config.log_trace_enabled:
        Log( level = 'TRACE', offset = offset ).emit( message )

def debug( message : str, offset : int = 0 ):
    if config.log_debug_enabled:
        Log( level = 'DEBUG', offset = offset ).emit( message )

def info( message : str, offset : int = 0 ):
    Log( level = 'INFO', offset = offset ).emit( message )

def alarm( message : str, offset : int = 0 ):
    Log( level = 'ALARM', offset = offset ).emit( message )

def event( message : str, offset : int = 0 ):
    Log( level = 'EVENT', offset = offset ).emit( message )
    

File: golang/log/log.go
package log

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"
)

type Log struct {
	Level     string
	Timestamp time.Time
	Process   string
	Function  string
	File      string
	Line      int
}

func NewLog(level string, offset int) *Log {
	var log Log
	log.Level = level
	log.Timestamp = time.Now().UTC()
	if config.PodName != "" {
		log.Process = config.PodName
	} else {
		log.Process = os.Args[0]
	}
	pc, file, line, _ := runtime.Caller(offset + 2)
	log.File = log.File[strings.LastIndex(file, "/")+1:]
	log.Line = line
	log.Function = runtime.FuncForPC(pc).Name()
	log.Function = log.Function[strings.LastIndex(log.Function, "/")+1:]
	return &log
}

func (l *Log) Emit(template string, args ...interface{}) {
	message := fmt.Sprintf(template, args...)
	fmt.Printf("%s | %s | %s | %s:%d | %s | %s\n", l.Level, l.Timestamp, l.Process, l.File, l.Line, l.Function, message)
}

func Trace(template string, args ...interface{}) {
	if config.TraceEnabled {
		NewLog("TRACE", 1).Emit(template, args...)
	}
}

func Debug(template string, args ...interface{}) {
	if config.DebugEnabled {
		NewLog("DEBUG", 1).Emit(template, args...)
	}
}

func Info(template string, args ...interface{}) {
	NewLog("INFO", 1).Emit(template, args...)
}

func Alarm(template string, args ...interface{}) {
	NewLog("ALARM", 1).Emit(template, args...)
}

func Event(template string, args ...interface{}) {
	NewLog("EVENT", 1).Emit(template, args...)
}
